### ISSUE TITLE
Specifying format style using command line option

### ISSUE THREAD

**Issue by ddovod** [clangd/clangd#26]
Hello guys. I've posted the same message to clangd-dev, just in case posting here.

Clangd by default tries to use .clang-format file to format code. Since "Format as you type" is experimental feature, it may be unusual for people like me who are used to format code before commit or using some keybindings. And personally I have some discomfort when my editor tries to do too much things when I'm typing.
Please consider adding a new command line option. Here's a patch which adds a new option to control formatting style. Using it I can pass "-clang-format-style={DisableFormat: true, SortIncludes: false}" and clangd will not apply any formatting to my code. Since there's "-clang-tidy-checks" for clang-tidy, it looks reasonable to have the same option for clang-format.
Thank you very much!

```
diff --git a/clang-tools-extra/clangd/tool/ClangdMain.cpp b/clang-tools-extra/clangd/tool/ClangdMain.cpp
index 53f7264e565..e73394f16c5 100644
--- a/clang-tools-extra/clangd/tool/ClangdMain.cpp
+++ b/clang-tools-extra/clangd/tool/ClangdMain.cpp
@@ -219,6 +219,12 @@ static llvm::cl::opt<bool> SuggestMissingIncludes(
                    "includes using index."),
     llvm::cl::init(true));
 
+static llvm::cl::opt<std::string> FormatStyle(
+    "clang-format-style",
+    llvm::cl::desc("Coding style to format code. "
+                   "Please refer to clang-format style option for details."),
+    llvm::cl::init(""));
+
 namespace {
 
 /// \brief Supports a test URI scheme with relaxed constraints for lit tests.
@@ -458,6 +464,9 @@ int main(int argc, char *argv[]) {
   }
   Opts.ClangTidyOptProvider = ClangTidyOptProvider.get();
   Opts.SuggestMissingIncludes = SuggestMissingIncludes;
+  if (!FormatStyle.empty()) {
+    clang::format::DefaultFormatStyle = FormatStyle.c_str();
+  }
   ClangdLSPServer LSPServer(
       *TransportLayer, FSProvider, CCOpts, CompileCommandsDirPath,
       /*UseDirBasedCDB=*/CompileArgsFrom == FilesystemCompileArgs, Opts);
```

**Comment by sam-mccall** [comment]
Sorry about the silence here.

I think being able to turn off format-as-you-type is important, and being able to control the style is important, but I wonder if connecting them is the right approach.

TL;DR: you can disable format-as-you-type on the editor side in vscode today.
I don't think `-style` is the right way to disable it, or even that useful, but we should support `-fallback-style`.

### Controlling format-as-you-type:
This is definitely an... invasive feature. And I'm not happy with the state of it (https://reviews.llvm.org/D60605)

But it's an LSP call the editor is specifically making, and clangd is unlikely to be the only language server where FAYT is jarring to some users. Editor preferences seem like a reasonable place to control this. Adding a flag that says "this call should fail/do nothing" seems like a last resort. And the proposed flag would also make explicit "format document" requests not work.

As far as I know, the only editor that implements format-as-you-type is VSCode. 
VSCode allows you to configure format-as-you-type using the `editor.formatOnType` setting. You can also enable this for particular languages (e.g. C++ only): https://code.visualstudio.com/docs/getstarted/settings#_language-specific-editor-settings

### Controlling formatting style:
`clang-format` supports two flags for styles: `-style` (always used) and `-fallback-style` (used when no `.clang-format` file is found).

One problem with controlling clangd behavior with a `-style` flag is that styles often differ between project, and your clangd flags are part of your editor config which tends to be global. This is why checking in `.clang-format` files is recommended. It's totally reasonable to want to change the default (like `-fallback-style`), but if `.clang-format` exists we should have a pretty good reason to ignore it.

IME the `-style` flag is mostly used for experimenting/testing clang-format, and for custom integrations. If there are other important use cases we can add it to clangd later.

Related: #10 (about disabling formatter).

**Comment by ddovod** [comment]
Hi @sam-mccall , thank you for detailed response!

Totally agree about mixing "turn off" and "style control" features. My patch is just the most simple solution I came up with, so if there will be an option to totally disable FAYT, it would be great.

I'm using emacs + lsp-mode (it has clangd integration out of the box), everything works great, at least I'm able to configure some features using clangd command line options globally. So far I'm using custom clangd build with patch from the first message. I'm not sure if lsp-mode have options to control FAYT, I'll check it.

Apart from that it looks reasonable to read clangd options from some dotfile, just like for clang-format and clang-tidy, and this idea is already somewhere in the `ClangdMain.cpp` file:
```
// FIXME: Flags are the wrong mechanism for user preferences.
// We should probably read a dotfile or similar.
```
Using dotfiles I would be able to configure needed behaviour on per-project basis.

And thank you very much for working on clangd!

**Comment by sam-mccall** [comment]
Ah, lsp-mode!

I'm not an emacs user, and at work the emacs folks are using eglot, so I didn't realize there was another editor that supported it.

Apparently `lsp-enable-on-type-formatting` is a setting you can tweak. Don't know emacs well enough to understand if that can be applied to certain filetypes only.

> Apart from that it looks reasonable to read clangd options from some dotfile, just like for clang-format and clang-tidy, and this idea is already somewhere in the ClangdMain.cpp file:
> Using dotfiles I would be able to configure needed behaviour on per-project basis.

Ah, IIRC that comment was referring to user preferences that are global across projects. (Like the -pch-storage or -completion-style flags).
A dotfile that's located near the code and stores project settings is an interesting idea, but if we're going to encourage users to check something like that in it needs to be designed carefully.

But I'm a bit confused - is something like "enable format-on-type" something you'd want different settings for different projects?

**Comment by ddovod** [comment]
Oh god, `lsp-enable-on-type-formatting` is exactly what I was looking for... I don't know what I was thinking about when started digging into clangd code to disable this feature, probably I thought it is clangd specific thing. Thank you very much!

About per-project settings, honestly I don't think it can be super helpful. I'm regularly working on 3-5 projects on my day-to-day work, and for me it's comfortable enough to have all clangd options enabled globally (right now it is `-header-insertion-decorators=false`, `-include-ineligible-results` and `-compile-commands-dir=<dir>`, the `<dir>` is evaluated from the elisp). But I've seen some bug reports or discussions in the mailing lists about clangd features that don't work well on some specific types of projects, so I know at least 2 cases when it could be useful to have per-project settings:

- compile-commands-dir - not very important since you can create a symlink to the cdb file, but anyway
- Automatic include fixer - it breaks some projects (don't remember why, probably some embedded things), however it works perfectly for all of the projects I'm working on.

What about this particular issue, now it's totally clear for me and can be closed. Thanks!

**Comment by sam-mccall** [comment]
Great! Yes, the header insertion is something we may need to support per-project customization. Early days though, we've been able to fix some of the broken cases.
I've added the `-fallback-style` flag in r360016, too.

**Comment by ddovod** [comment]
Great, thank you very much!
